r/R/singlecell_utils.R

```diff
@@ -68,4 +68,201 @@ marker_features <- function(mat, groups, method="wilcoxon") {
+#'
+#' Current options are: `nonzeros`, `sum`, `mean`, `variance`.
+#' @param clip_values (integer) If set, clip matrix values to the `<clip_values>`th quantile of each cell, between 0 and 1 inclusive.
+#' @param threads (integer) Number of threads to use.
+#' @return
+#'  - If `method` is length `1`, returns a matrix of shape `(features x groups)`.
+#'  - If `method` is greater than length `1`, returns A list of matrices with each matrix representing a pseudobulk matrix with a different aggregation method.
+#' Each matrix is of shape `(features x groups)`, and names are one of `nonzeros`, `sum`, `mean`, `variance`.
+#' @details The stats are ordered by complexity: nonzero, sum, mean, then variance. All
+#' less complex stats are calculated in the process of calculating a more complicated stat.
+#' That means it will not take additional time to calculate `mean` if `var` is also requested.
+#' @inheritParams marker_features
+#' @export
+pseudobulk_matrix <- function(mat, cell_groups, method = "sum", clip_values = 1, threads = 1L) {
+  assert_is(mat, "IterableMatrix")
+  assert_is(cell_groups, c("factor", "character", "numeric"))
+  cell_groups <- as.factor(cell_groups)
+  assert_is(method, "character")
+  methods <- c("variance", "mean", "sum", "nonzeros")
+  for (m in method) {
+    match.arg(m, choices = methods)
+  }
+  assert_is(clip_values, "numeric")
+  assert_true(all(clip_values >= 0 & clip_values <= 1))
+  assert_is(threads, "integer")
+  # if multiple methods are provided, only need to pass in the top method as it will also calculate the less complex stats
+  if (clip_values != 1) {
+    if (mat@transpose) {
+      quantile_values <- rowQuantiles(mat, probs = clip_values, type = 7L)
+    } else {
+      quantile_values <- colQuantiles(mat, probs = clip_values, type = 7L)
+    }
+    # if all values are 0, skip
+    mat <- min_by_col(mat, quantile_values)
+  }
+  iter <- iterate_matrix(parallel_split(mat, threads, threads*4))
+  cell_groups <- as.factor(cell_groups)
+  res <- pseudobulk_matrix_cpp(iter, cell_groups = as.integer(cell_groups) - 1, method = method, transpose = mat@transpose)
+  # if res is a single matrix, return with colnames and rownames
+  if (length(method) == 1) {
+    colnames(res[[method]]) <- levels(cell_groups)
+    rownames(res[[method]]) <- rownames(mat)
+    return(res[[method]])
+  }
+  # give colnames and rownames for each matrix in res, which is a named list
+  for (res_slot in names(res)) {
+    if ((length(res[[res_slot]]) == 0) || !(res_slot %in% method)) {
+      res[[res_slot]] <- NULL
+    } else {
+      colnames(res[[res_slot]]) <- levels(cell_groups)
+      rownames(res[[res_slot]]) <- rownames(mat)
+    }
+  }
+  return(res)
+}
+
+
+#' Find the nth quantile value(s) of each row in a matrix. Only supports transposed matrices.
```

Comment:
I noticed that I can't add params in, or else the `IterableMatrix methods` page gets all screwed up.  It is probably a good time to start splitting that page up into a new section